dm/syncer/checkpoint.go

```diff
@@ -517,34 +622,43 @@ func (cp *RemoteCheckPoint) FlushPointsExcept(tctx *tcontext.Context, exceptTabl
 	sqls := make([]string, 0, 100)
 	args := make([][]interface{}, 0, 100)
 
-	if cp.globalPoint.outOfDate() || cp.globalPointSaveTime.IsZero() || cp.needFlushSafeModeExitPoint {
-		locationG := cp.GlobalPoint()
+	if cp.globalPoint.outOfDateBy(snapshotCp.globalPoint.location) || cp.globalPointSaveTime.IsZero() || snapshotCp.needFlushSafeModeExitPoint {
+		locationG := snapshotCp.globalPoint.location
 		sqlG, argG := cp.genUpdateSQL(globalCpSchema, globalCpTable, locationG, cp.safeModeExitPoint, nil, true)
```

Comment:
Plan to fix this issue, let me know if sounds good to you:
- use cp.needFlushSafeModeExitPoint instead of snapshotCp.needFlushSafeModeExitPoint here
- since flush safe mode exit point only happen for sync flush method, we only check if the current flush is sync flush
- avoid to pass `needFlushSafeModeExitPoint `, instead we only add global point as part of snapshot when it pass check in `Snapshot()` creation method

Basically, these actions could simplify the mental model of flushing safe mode exit point here. safe mode exit point will only be flushed synchronously, and it only happen sequentially, so the worst case is we flush it twice(still doubt if that could ever happen), and we are also safe to flush it twice as it's done that incrementally.

Reply:
> since flush safe mode exit point only happen for sync flush method, we only check if the current flush is sync flush

our old logic abut "safe mode exit point" is when the replication progress pass the point, we are going to clear "safe mode exit point"  in downstream checkpoint. Now replication progress will fire two types of flush: sync flush by binlog heartbeat event and async flush when `CheckGlobalPoint && CheckLastSnapshotCreationTime`. I'm double checking that you mean we only let sync flush update "safe mode exit point"?

If so I think it's OK, since global checkpoint is forwarded, an outdated "safe mode exit point" should not be hurt.

> use cp.needFlushSafeModeExitPoint instead of snapshotCp.needFlushSafeModeExitPoint here ....

I agree we use current checkpint rather than snapshot checkpoint when dealing with "safe mode exit point".